lib/Doctrine/ORM/Internal/Hydration/AbstractHydrator.php

```diff
@@ -396,94 +410,173 @@ abstract protected function hydrateAllData();
      *                                             row, grouped by their
      *                                             component alias.
      * @psalm-return array{
+     *                   ids: array<array-key, array<array-key, mixed>>,
      *                   data: array<array-key, array>,
      *                   newObjects?: array<array-key, array{
      *                       class: mixed,
-     *                       args?: array
+     *                       args: array
      *                   }>,
      *                   scalars?: array
      *               }
      */
     protected function gatherRowData(array $data, array &$id, array &$nonemptyComponents)
     {
-        $rowData = ['data' => []];
+        $rowData = [
+            'ids'  => $this->gatherRowIDs($data, $id, $nonemptyComponents),
+            'data' => [],
+        ];
 
-        foreach ($data as $key => $value) {
-            $cacheKeyInfo = $this->hydrateColumnInfo($key);
-            if ($cacheKeyInfo === null) {
-                continue;
+        foreach (array_keys($this->_rowMappingInfo['dql']) as $alias) {
+            $rowData['data'][$alias] = $this->gatherRowDataFromDQLMapping($data, $alias);
+        }
+
+        if (isset($this->_rowMappingInfo['scalars'])) {
+            $rowData['scalars'] = $this->gatherRowDataFromScalarsMapping($data);
+        }
+
+        if (isset($this->_rowMappingInfo['newObjects'])) {
+            $rowData['newObjects'] = $this->gatherRowDataFromNewObjectsMapping($data);
+        }
+
+        return $rowData;
+    }
+
+    /**
+     * @param mixed[] $data SQL Result Row.
+     * @psalm-param array<array-key, string> $id                 Dql-Alias => ID-Hash.
+     * @psalm-param array<array-key, bool>   $nonemptyComponents Does this DQL-Alias has at least one non NULL value?
```

Comment:
```suggestion
     * @param string[] $id                 Dql-Alias => ID-Hash.
     * @param bool[]   $nonemptyComponents Does this DQL-Alias has at least one non NULL value?
```

No need to vendor-prefix those.